### PR TITLE
fix: Fix non garbage collect on response sender

### PR DESCRIPTION
**Pull request by kthui**
#### What does the PR do?

Response sender(s) may not be deleted and garbage collected properly when it is passed to another thread. This PR ensures it is deleted when it goes out of scope.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [ ] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [ ] Added [test plan](#test-plan) and verified test passes.
- [ ] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [ ] feat
- [x] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Related PRs:

https://github.com/triton-inference-server/vllm_backend/pull/46

#### Where should the reviewer start?

Start with the [related PR](https://github.com/triton-inference-server/vllm_backend/pull/46) and then look into this fix and some minor refactoring.

#### Test plan:

Existing tests on vLLM backend should be sufficient at covering any regression bring by the performance improvement (and fix).

- CI Pipeline ID:

#### Caveats:

N/A

#### Background

N/A

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
- closes GitHub issue: N/A

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -300,6 +280,28 @@ def get_sampling_params_dict(self, params_json):
...
+            except Exception as e:
+                self.logger.log_error(
+                    f"An error occurred while sending a response: {e}"
+                )
+            finally:
+                if response_flag == pb_utils.TRITONSERVER_RESPONSE_COMPLETE_FINAL:
+                    self.ongoing_request_count -= 1
+                    del response_sender
+                    if self._response_queue.empty():
+                        gc.collect()
```
Comment (Tabrizian) [inline src/model.py:302]: Can we do garbage collection on `None` item only? I'm a bit concerned with calling garbage collection for every request completion.

Comment (kthui) [reply src/model.py:302]: > Can we do garbage collection on `None` item only?

The `None` item is only placed into the response queue when the `finalize()` function is called, but the `finalize()` function is called only if all the response sender(s) are destructed. In other words, we will only see the `None` item after the `response_sender`(s) are garbage collected.

I think the best solution now is to garbage collect if the response queue is empty, which means garbage collect happens only if there is no new response to be sent at the moment a request is completed, so it should not create too much overhead ideally.

Here is the frequency and duration of garbage collects happened during benchmarking:
```
I0722 21:40:56.330660 2595 grpc_server.cc:2463] "Started GRPCInferenceService at 0.0.0.0:8001"
I0722 21:40:56.330826 2595 http_server.cc:4692] "Started HTTPService at 0.0.0.0:8000"
I0722 21:40:56.383418 2595 http_server.cc:362] "Started Metrics Service at 0.0.0.0:8002"
I0722 21:41:30.977492 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:41:31.072659 2595 model.py:304] "----- Garabage collect end -----"
I0722 21:41:55.523516 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:41:55.618196 2595 model.py:304] "----- Garabage collect end -----"
I0722 21:42:20.098470 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:42:20.193094 2595 model.py:304] "----- Garabage collect end -----"
I0722 21:42:44.672155 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:42:44.766406 2595 model.py:304] "----- Garabage collect end -----"
I0722 21:43:09.278387 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:43:09.372832 2595 model.py:304] "----- Garabage collect end -----"
I0722 21:43:33.794605 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:43:33.889078 2595 model.py:304] "----- Garabage collect end -----"
I0722 21:43:58.329119 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:43:58.422977 2595 model.py:304] "----- Garabage collect end -----"
I0722 21:44:22.812579 2595 model.py:302] "----- Garabage collect start -----"
I0722 21:44:22.906684 2595 model.py:304] "----- Garabage collect end -----"
```

**Review by Tabrizian — APPROVED** [review]
We need to fix the lifetime issue of finalize not getting called when the response factory is not deleted in the core.

**Comment by GuanLuo** [comment]
Why response sender needs to be gc'ed right the way?

**Comment by kthui** [comment]
> Why response sender needs to be gc'ed right the way?

Because garbage collecting the response sender will call its destructor on the C++ side, which then calls the destructor on the response factory which holds a shared pointer to the model. The shared pointer will stop the model from gracefully unloading.

We are not garbage collecting right away, it will only collect based on a heuristic that determines if there is no more ongoing request, which generally only collect once at the end of a busy benchmarking.

**Pull request merged**